sqrt/sqrtdecomposition.go

```diff
@@ -0,0 +1,102 @@
+// Package sqrt contains algorithms and data structures that contains a √n in their complexity
+package sqrt
+
+import "math"
+
+// Sqrt (or Square Root) Decomposition is a technique used for query an array and perform updates
+// Inside this package is described its most simple data structure, you can find more at: https://cp-algorithms.com/data_structures/sqrt_decomposition.html
+//
+// Formally, You can use SqrtDecomposition only if:
+//
+// Given a function $Query:E_1,...,E_n\rightarrow Q$
+//
+// if $\exist unionQ:Q,Q\rightarrow Q$
+//
+// s.t.
+//
+// - $\forall n\in \N > 1, 1\le i<n, E_1,..., E_n\in E \\ query(E_1,..., E_n)=unionQ(query(E_1,..., E_i), query(E_{i+1},...,E_n))$
+//
+// - (Only if you want use $update$ function)
+// $\forall n\in \N > 0, E_1,..., E_n\in E \\ query(E_1,...,E_{new},..., E_n)=updateQ(query(E_1,...,E_{old},...,E_n), indexof(E_{old}), E_{new})$
+type SqrtDecomposition[E any, Q any] struct {
+	querySingleElement func(element E) Q
+	unionQ             func(q1 Q, q2 Q) Q
+	updateQ            func(oldQ Q, oldE E, newE E) (newQ Q)
+
+	elements  []E
+	blocks    []Q
+	blockSize uint64
+}
+
+// Create a new SqrtDecomposition instance with the parameters as specified by SqrtDecomposition comment
+// Assumptions:
+//   - len(elements) > 0
+func NewSqrtDecomposition[E any, Q any](
```

Comment:
```suggestion
func NewDecomposition[E any, Q any](
```

Again sqrt is in the name

And maybe call this `Decomp`, instead of `Decomposition`